core/src/main/kotlin/org/evomaster/core/problem/rest/classifier/ModelAccuracyWithTimeWindow.kt

```diff
@@ -19,7 +22,29 @@ class ModelAccuracyWithTimeWindow(
         return ok / n
     }
 
-    override fun updatePerformance(predictionWasCorrect: Boolean) {
+    /**
+     * Precision for HTTP 400 responses: TP / (TP + FP).
+     */
+    override fun estimatePrecision400(): Double {
+        val tp = truePositive400Queue.count { it }   // count only true entries
+        val fp = falsePositive400Queue.count { it }  // count only true entries
+        return if ((tp + fp) > 0) tp.toDouble() / (tp + fp) else 0.0
+    }
+
+    /**
+     * Update performance based on prediction vs actual result.
+     */
+    override fun updatePerformance(predictedStatusCode: Int, actualStatusCode: Int?) {
+
+        val predictionWasCorrect = predictedStatusCode == actualStatusCode
+
         queue.add(predictionWasCorrect)
+
+        if (actualStatusCode == 400) {
+            truePositive400Queue.add(predictionWasCorrect)
+        } else {
+            falsePositive400Queue.add(!predictionWasCorrect)
+        }
```

Comment:
is this correct? at each call of `updatePerformance`, should both data structures `truePositive400Queue` and `falsePositive400Queue` be updated?  regardless, add unit tests to validate the precision calculations